controllers/user_controller.go

```diff
@@ -0,0 +1,428 @@
+					fmt.Sprintf("Generation of k8s secrets [%s] has failed", secret.Name), err)
+			}
+			// Secret created successfully - requeue only if no version is specified
+			return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorSucceeded", metav1.ConditionTrue, "SecretCreationSuccess",
+				fmt.Sprintf("The secret [%s] was created according to its matching custom resource", secret.Name), nil)
+		}
+	}
+
+	if !controllerutil.ContainsFinalizer(userResource, userFinalizer) {
+		controllerutil.AddFinalizer(userResource, userFinalizer)
+		err := r.Update(ctx, userResource)
+		if err != nil {
+			logger.Error(err, "Failed to add finalizer.")
+			return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorFailed", metav1.ConditionFalse, "S3UserResourceCreationFailed",
+				fmt.Sprintf("Failed to add finalizer on userResource [%s] has failed", userResource.Name), err)
+		}
+	}
+
+	return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorSucceeded", metav1.ConditionTrue, "S3UserCreated",
+		fmt.Sprintf("The user [%s] was created", userResource.Spec.Name), nil)
+}
+
+// SetupWithManager sets up the controller with the Manager.*
+func (r *S3UserReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&s3v1alpha1.S3User{}).
+		Owns(&corev1.Secret{}).
+		// TODO : implement a real strategy for event filtering ; for now just using the example from OpSDK doc
+		// (https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/)
+		WithEventFilter(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// Ignore updates to CR status in which case metadata.Generation does not change
+				return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				// Evaluates to false if the object has been confirmed deleted.
+				return !e.DeleteStateUnknown
+			},
+		}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Complete(r)
+}
+
+func (r *S3UserReconciler) setS3UserStatusConditionAndUpdate(ctx context.Context, userResource *s3v1alpha1.S3User, conditionType string, status metav1.ConditionStatus, reason string, message string, srcError error) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	meta.SetStatusCondition(&userResource.Status.Conditions,
+		metav1.Condition{
+			Type:               conditionType,
+			Status:             status,
+			Reason:             reason,
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Message:            message,
+			ObservedGeneration: userResource.GetGeneration(),
+		})
+
+	err := r.Status().Update(ctx, userResource)
+	if err != nil {
+		logger.Error(err, "an error occurred while updating the status of the user resource")
+		return ctrl.Result{}, utilerrors.NewAggregate([]error{err, srcError})
+	}
+	return ctrl.Result{}, srcError
+}
```

Comment:
In the last version released for s3-operator, we moved away from meta.SetStatusCondition because of the behavior for lastTransitionTime was not convenient for us. Could you reimplement this accordingly - you can probably just copy/paste/rename the relevant part from another controller, eg : https://github.com/InseeFrLab/s3-operator/blob/main/controllers/bucket_controller.go#L221-L267